### PR TITLE
feat: slow sync calculates supported protocols [WPB-6660]

### PR DESCRIPTION
**Pull request by markusfassbender**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-6660" title="WPB-6660" target="_blank">Sub-taskWPB-6660</a>  [iOS] calculate supported protocols during slow sync
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

💡 this merges into epic branch #944

- see ticket

### Solutions

Implements new slow sync phase to supported calculate protocols.

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

- set break points for the slow sync phases and validate the right order of execution

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-request-strategy/Tests/Sources/Services/SupportedProtocolsServiceTests.swift [hunk 1]
```diff
@@ -16,47 +16,62 @@
...
 //
 
-import Foundation
 import XCTest
-@testable import WireSyncEngine
 import WireDataModelSupport
+@testable import WireRequestStrategy
 
-final class SupportedProtocolsServiceTests: MessagingTest {
+final class SupportedProtocolsServiceTests: XCTestCase {
```
Comment (markusfassbender) [inline wire-ios-request-strategy/Tests/Sources/Services/SupportedProtocolsServiceTests.swift:23]: This was needed to move the test in `WireRequestStrategy`

File: wire-ios-request-strategy/WireRequestStrategy.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -450,7 +455,6 @@
 		0605DB8F2511622100443219 /* ZMLocalNotificationTests_UnreadCount.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = ZMLocalNotificationTests_UnreadCount.swift; sourceTree = "<group>"; };
 		060ED6D52499F41000412C4A /* PushNotificationStatus.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = PushNotificationStatus.swift; sourceTree = "<group>"; };
 		062285C7276BB5B000B87C91 /* UpdateEventProcessor.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = UpdateEventProcessor.swift; sourceTree = "<group>"; };
-		062DD2D72760CFAA00E27FD9 /* UUID+SafeLogging.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "UUID+SafeLogging.swift"; sourceTree = "<group>"; };
```
Comment (markusfassbender) [inline wire-ios-request-strategy/WireRequestStrategy.xcodeproj/project.pbxproj:453]: This was a dead file reference

**Review by caldrian — APPROVED** [review]

File: wire-ios-request-strategy/Sources/Request Strategies/User/SelfSupportedProtocolsRequestStrategy.swift [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public final class SelfSupportedProtocolsRequestStrategy: AbstractRequestStrategy, ZMSingleRequestTranscoder {
+
+    // Slow Sync
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Request Strategies/User/SelfSupportedProtocolsRequestStrategy.swift:25]:
```suggestion
    // MARK: - Slow Sync
```

File: wire-ios-request-strategy/Sources/Request Strategies/User/SelfSupportedProtocolsRequestStrategy.swift [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    private unowned var syncProgress: SyncProgress
+
+    private let syncPhase: SyncPhase = .updateSelfSupportedProtocols
+
+    private var isSlowSyncing: Bool { syncProgress.currentSyncPhase == syncPhase }
+
+    // Requests
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Request Strategies/User/SelfSupportedProtocolsRequestStrategy.swift:33]:
```suggestion
    // MARK: - Requests
```

File: wire-ios-request-strategy/Sources/Request Strategies/User/SelfSupportedProtocolsRequestStrategy.swift [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+        configuration = [.allowsRequestsDuringSlowSync]
+    }
+
+    public override func nextRequestIfAllowed(for apiVersion: APIVersion) -> ZMTransportRequest? {
+        guard isSlowSyncing else {
+            return nil
+        }
+
+        WireLogger.sync.info("start slow sync phase: \(syncPhase.description)")
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Request Strategies/User/SelfSupportedProtocolsRequestStrategy.swift:62]: non-blocking: we should unify how we log sync phase. but one thing to know, there's a log in SyncStatus that will log the syncPhase when set producing a log on datadog.

It's not log on prod device though

File: wire-ios-request-strategy/Sources/Request Strategies/User/SelfSupportedProtocolsRequestStrategy.swift [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        guard isSlowSyncing else {
+            // skip result if we are not in the slow sync...
+            assertionFailure("expected response during slow sync phase!")
+            WireLogger.sync.error("received response, but expected during slow sync phase '\(syncPhase.description)'!")
+            return
+        }
+
+        switch response.result {
+        case .success:
+            WireLogger.sync.error("finished slow sync phase '\(syncPhase.description)'!")
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Request Strategies/User/SelfSupportedProtocolsRequestStrategy.swift]:
```suggestion
            WireLogger.sync.info("finished slow sync phase '\(syncPhase.description)'!")
```

File: wire-ios-request-strategy/Tests/Sources/Request Strategies/User/SelfSupportedProtocolsRequestBuilderTests.swift [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import XCTest
+@testable import WireRequestStrategy
+
+final class SelfSupportedProtocolsRequestBuilderTests: XCTestCase {
+
+    // the api version is just required to build and not influence the tests themselves
+    private let defaultAPIVersion: APIVersion = .v4
```
Comment (netbe) [inline wire-ios-request-strategy/Tests/Sources/Request Strategies/User/SelfSupportedProtocolsRequestBuilderTests.swift]: isn't it apiVersion v5?

Comment (netbe) [reply wire-ios-request-strategy/Tests/Sources/Request Strategies/User/SelfSupportedProtocolsRequestBuilderTests.swift]: suggestion: change to `minimumAPIVersion`

File: wire-ios-request-strategy/Tests/Sources/Request Strategies/User/SelfSupportedProtocolsRequestBuilderTests.swift [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        let request = builder.buildUpstreamRequest(keys: .init())
+
+        // then
+        XCTAssert(request?.userInfo.isEmpty == true)
+    }
+
+    // MARK: Helpers
+
+    private func makeBuilder(
+        apiVersion: APIVersion? = nil,
```
Comment (netbe) [inline wire-ios-request-strategy/Tests/Sources/Request Strategies/User/SelfSupportedProtocolsRequestBuilderTests.swift:181]: suggestion (non-blocking): make apiVersion non optional but with default value of the defaultApiVersion

**Review by netbe — APPROVED** [review]
LGTM;) just some minor things about logging and some questions

File: wire-ios-request-strategy/Tests/Sources/Request Strategies/User/SelfSupportedProtocolsRequestBuilderTests.swift (hunk 8 above)
Comment (markusfassbender) [reply wire-ios-request-strategy/Tests/Sources/Request Strategies/User/SelfSupportedProtocolsRequestBuilderTests.swift:181]: Unfortunately it's not possible to reference the var `minimumAPIVersion` as default argument here. That's why I chose this pattern to use `nil` and internally fall back to use the default.

File: wire-ios-request-strategy/Sources/Request Strategies/User/SelfSupportedProtocolsRequestStrategy.swift (hunk 5 above)
Comment (markusfassbender) [reply wire-ios-request-strategy/Sources/Request Strategies/User/SelfSupportedProtocolsRequestStrategy.swift:62]: I agree, though would not change it here in this PR now. Thanks!

**Pull request merged**